IBX-323 - Extending Solr (#1369)
ezsystems/developer-documentation: docs/guide/search/extend_solr.md

- `EzSystems\EzPlatformSolrSearchEngine\Query\Common\AggregationVisitor\TermAggregationVisitor`
- `EzSystems\EzPlatformSolrSearchEngine\ResultExtractor\AggregationResultExtractor\StatsAggregationResultExtractor`
- `EzSystems\EzPlatformSolrSearchEngine\ResultExtractor\AggregationResultExtractor\TermAggregationResultExtractor`

In a more complex use case, you must create your own visitor and extractor.

### Custom aggregation visitor

The aggregation visitor must implement `EzSystems\EzPlatformSolrSearchEngine\Query\AggregationVisitor`:

``` php
[[= include_file('code_samples/front/search_extensibility/solr/src/Query/Aggregation/PriorityRangeAggregationVisitor.php') =]]
```

The `canVisit()` method checks whether the provided aggregation is of the supported type
(in this case, your custom `PriorityRangeAggregation`).

The `visit()` method returns an array of results.

### Custom result extractor

You must also create a result extractor, which implements  `EzSystems\EzPlatformSolrSearchEngine\ResultExtractor\AggregationResultExtractor` 
that transforms raw aggregation results from Solr into `AggregationResult` objects:

``` php

> [[= include_file('code_samples/front/search_extensibility/solr/src/Query/Aggregation/PriorityRangeAggregationExtractor.php') =]]

Author: ```suggestion
[[= include_file('code_samples/front/search_extensibility/solr/src/Query/Aggregation/PriorityAggregationResultExtractor.php') =]]
```

# Solr extensibility

[Solr](solr.md) can be extended by adding several functionalities, such as document field mappers, custom search criteria, custom sort clauses, or custom aggregations.

## Document field mappers

!!! note

>     Document field mappers are available since Solr bundle version 1.2.

Reviewer: I agree with @alongosz here.

- `LocationFieldMapper`
    - `::accept(Location $content)`
    - `::mapFields(Location $content)`

Mappers can be used on the extension points by registering them with the [service container](../service_container.md) by using service tags, as follows:

- All block documents
    - `ContentFieldMapper`
    - `ezpublish.search.solr.field_mapper.block`
- All block documents per translation
    - `ContentTranslationFieldMapper`
    - `ezpublish.search.solr.field_mapper.block_translation`
- Content documents
    - `ContentFieldMapper`
    - `ezpublish.search.solr.field_mapper.content`
- Content documents per translation
    - `ContentTranslationFieldMapper`
    - `ezpublish.search.solr.field_mapper.content_translation`
- Location documents
    - `LocationFieldMapper`
    - `ezpublish.search.solr.field_mapper.location`

The following example shows how you can index data from the parent Location content, 
to make it available for full-text search on the child content.
The example relies on a use case of indexing webinar data on the webinar events, 
which are children of the webinar. 
The field mapper could then look like this:

```php
 <?php

> namespace My\WebinarApp;

Reviewer 2: Those examples are inconsistent. The next one has `App` namespace, while this one has `My` namespace. More something like
```suggestion
namespace App\Search\Mapper;
```

        return [
            new Search\Field(
                'fulltext',
                $parentContentInfo->name,
                new Search\FieldType\FullTextField()
            ),
        ];
    }
}
```

You index full text data only on the content document, therefore, you would register the service like this:

``` yaml
my_webinar_app.webinar_event_title_fulltext_field_mapper:
    class: My\WebinarApp\WebinarEventTitleFulltextFieldMapper
    arguments:
        - '@ezpublish.spi.persistence.content_handler'
        - '@ezpublish.spi.persistence.location_handler'
    tags:
        - {name: ezpublish.search.solr.field_mapper.content}
```

!!! caution "Permission issues when using Repository API in document field mappers"

    Document field mappers are low-level and expect to be able to index all content 
    regardless of current user permissions.
    If you use PHP API in your custom document field mappers, apply [`sudo()`](../../api/public_php_api.md#using-sudo),
    or use the Persistence SPI layer as in the example above.

## Custom Search Criteria

To provide support for a custom Search Criterion, do the following.

> First, in the `src/Query/Criterion/CameraManufacturerVisitor.php` file, implement `CriterionVisitor`:

Reviewer 2: I disagree on the order here. First you should create a Criterion and then a Visitor for it, since the Visitor depends on the criterion.

``` php
<?php

declare(strict_types=1);

namespace App\Query\Criterion;

use eZ\Publish\API\Repository\Values\Content\Query\Criterion;
use EzSystems\EzPlatformSolrSearchEngine\Query\CriterionVisitor;

final class CameraManufacturerVisitor extends CriterionVisitor
{
    public function canVisit(Criterion $criterion)
    {
        return $criterion instanceof CameraManufacturer;
    }
    public function visit(Criterion $criterion, CriterionVisitor $subVisitor = null)
    {
        $expressions = array_map(
            static function ($value): string {
                return 'exif_camera_manufacturer_id:"' . $this->escapeQuote($value) . '"';
            },

>             ($criterion->value

Reviewer 2: Looks like a typo
```suggestion
            $criterion->value
```

final class CameraManufacturer extends Criterion
{
    /**
     * @param string|string[] $value Manufacturer name(s) to be matched.
     */
    public function __construct($value)
    {
        parent::__construct(null, null, $value);
    }

    public function getSpecifications(): array
    {
        return [
            new Specifications(
                Operator::IN,
                Specifications::FORMAT_ARRAY,
                Specifications::TYPE_STRING
            ),
            new Specifications(
                Operator::EQ,
                Specifications::FORMAT_SINGLE,
                Specifications::TYPE_STRING
            ),
        ];
    }
}
```

Finally, register the visitor as a service.

Search Criteria can be valid for both Content and Location search.
To choose the search type, use either `content` or `location` in the tag:

``` yaml
services:
    App\Query\Criterion\CameraManufacturerVisitor:            
        tags:
            - { name: ezpublish.search.solr.query.content.criterion_visitor }
            - { name: ezpublish.search.solr.query.location.criterion_visitor }    
```

## Custom Sort Clause

To create a custom Sort Clause, do the following.

> First, in the `src/Query/SortClause/ScoreVisitor.php` file, implement `SortClauseVisitor`:

Reviewer 2: Again, the order should be different. SortClause first, followed by its Visitor.

declare(strict_types=1);

namespace App\Query\SortClause;

use eZ\Publish\API\Repository\Values\Content\Query;
use eZ\Publish\API\Repository\Values\Content\Query\SortClause;

final class Score extends SortClause
{
    public function __construct(string $sortDirection = Query::SORT_ASC)
    {
        parent::__construct('_score', $sortDirection);
    }
}
```

> Then, in the `src/Query/SortClause/ScoreVisitor.php` file, implement `SortClauseVisitor`:

Reviewer 3: ```suggestion
Then, add an `src/Query/SortClause/ScoreVisitor.php` file that implements `SortClauseVisitor`:
```

        tags:
            - { name: ezplatform.search.solr.query.location.aggregation_result_extractor } 
```

You can use a different type of aggregation, followed by respective visitor and extractor classes:

- `EzSystems\EzPlatformSolrSearchEngine\Query\Common\AggregationVisitor\StatsAggregationVisitor`
- `EzSystems\EzPlatformSolrSearchEngine\Query\Common\AggregationVisitor\TermAggregationVisitor`
- `EzSystems\EzPlatformSolrSearchEngine\ResultExtractor\AggregationResultExtractor\StatsAggregationResultExtractor`
- `EzSystems\EzPlatformSolrSearchEngine\ResultExtractor\AggregationResultExtractor\TermAggregationResultExtractor`

In a more complex use case, you must create your own visitor and extractor.

### Custom aggregation visitor

The aggregation visitor must implement `EzSystems\EzPlatformSolrSearchEngine\Query\AggregationVisitor`:

``` php
<?php

declare(strict_types=1);

namespace App\Query\Aggregation;
use eZ\Publish\API\Repository\Values\Content\Query\Aggregation;
use EzSystems\EzPlatformSolrSearchEngine\Query\AggregationVisitor;

final class PriorityRangeAggregationVisitor implements AggregationVisitor
{

>     public function canVisit(Aggregation $aggregation, array $languageFilter): bool

Author: @alongosz @adamwojs Is it OK to use  $languageFilter here as a parameter if we have not declared the use of
`Ibexa\Platform\Contracts\ElasticSearchEngine\Query\LanguageFilter` ?

Reviewer 2: > @alongosz @adamwojs Is it OK to use $languageFilter here as a parameter if we have not declared the use of
> `Ibexa\Platform\Contracts\ElasticSearchEngine\Query\LanguageFilter` ?

How is `LanguageFilter` related to `array`?

Author: Dunno, but parameter name caught my attention.

Reviewer 2: > Dunno, but parameter name that's been used in the code sample caught my attention.

It doesn't seem related.

    {
        return $aggregation instanceof PriorityRangeAggregation;
    }

>     /**

Reviewer 2: CS
```suggestion

    /**
```

     * @param \eZ\Publish\API\Repository\Values\Content\Query\Aggregation\AbstractRangeAggregation $aggregation
     */
    public function visit(
        AggregationVisitor $dispatcherVisitor,
        Aggregation $aggregation,
        array $languageFilter
    ): array {
        $rangeFacets = [];
        foreach ($aggregation->getRanges() as $range) {
            $from = $this->formatRangeValue($range->getFrom());
            $to = $this->formatRangeValue($range->getTo());
            $rangeFacets["${from}_${to}"] = [
                'type' => 'query',
                'q' => sprintf('priority_i:[%s TO %s}', $from, $to),
            ];
        }

>         return [

Reviewer 2: ```suggestion

        return [
```

            'type' => 'query',
            'q' => '*:*',
            'facet' => $rangeFacets,
        ];
    }

>     private function formatRangeValue($value): string

Reviewer 2: ```suggestion

    private function formatRangeValue($value): string
```

    {
        if ($value === null) {
            return '*';
        }

>         return (string)$value;

Reviewer 2: ```suggestion

        return (string)$value;
```

that transforms raw aggregation results from Solr into `AggregationResult` objects:

``` php
<?php

declare(strict_types=1);

namespace App\Query\Aggregation;

use eZ\Publish\API\Repository\Values\Content\Query\Aggregation;
use eZ\Publish\API\Repository\Values\Content\Query\Aggregation\Range;
use eZ\Publish\API\Repository\Values\Content\Search\AggregationResult;
use eZ\Publish\API\Repository\Values\Content\Search\AggregationResult\RangeAggregationResult;
use eZ\Publish\API\Repository\Values\Content\Search\AggregationResult\RangeAggregationResultEntry;
use EzSystems\EzPlatformSolrSearchEngine\ResultExtractor\AggregationResultExtractor;
use stdClass;

final class PriorityAggregationResultExtractor implements AggregationResultExtractor
{
    public function canVisit(Aggregation $aggregation, array $languageFilter): bool
    {
        return $aggregation instanceof PriorityRangeAggregation;
    }
    public function extract(Aggregation $aggregation, array $languageFilter, stdClass $data): AggregationResult
    {
        $entries = [];
        foreach ($data as $key => $bucket) {

>             if ($key === 'count') {
>                 continue;
>             }
>             if (strpos($key, '_') === false) {
>                 continue;
>             }

Reviewer 2: ```suggestion
            if ($key === 'count' || strpos($key, '_') === false) {
                continue;
            }
```

>             list($from, $to) = explode('_', $key, 2);

Reviewer 2: legacy practice (PHP <7.1)
```suggestion
            [$from, $to] = explode('_', $key, 2);
```

            $entries[] = new RangeAggregationResultEntry(
                new Range(
                    $from !== '*' ? $from : null,
                    $to !== '*' ? $to : null
                ),
                $bucket->count
            );
        }

>         return new RangeAggregationResult($aggregation->getName(), $entries);

Reviewer 2: Cs
```suggestion

        return new RangeAggregationResult($aggregation->getName(), $entries);
```

The field mapper could then look like this:

```php

> [[= include_file('code_samples/front/search_extensibility/solr/src/FieldMapper/WebinarEventTitleFulltextFieldMapper.php') =]]

Reviewer 3: `code_samples/front/` was supposed to cover rendering on the front page in general, here maybe `code_samples/search/solr/...`?

```

You index full text data only on the content document, therefore, you would register the service like this:

``` yaml

> [[= include_file('code_samples/front/search_extensibility/solr/src/FieldMapper/services.yaml') =]]

Reviewer 3: Isn't it `config/packages/services.yaml`? That's the regular location for the services file, AFAIK.

and the result extractor with `ezplatform.search.solr.query.location.aggregation_result_extractor`:

``` yaml

> [[= include_file('code_samples/front/search_extensibility/solr/src/Query/Aggregation/services.yaml', 19, 26) =]]

Reviewer 3: as above
